[RELAND] Disallow skipping dynamo

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110222

Previous discussion: https://github.com/pytorch/pytorch/pull/109476

In this PR, I made following additions to the original PR:
1) Unlifted graph module now runs the runtime assertions in its' forward call. 
2) When we retrace, we make sure we run the assertions to make sure user is tracing the module with correct inputs with respect to the assumptions we made during first tracing. The way I do is that I create new graph module type with modified call method. And the runtime assertions happen under torchdynamo.disable so that it is just run in eager directly. The reason is we don't this to be traced part of the graph. 
3) Both ep.module and capture_pre_autograd now returns _UnliftedGraphModule. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @avikchaudhuri @gmagogsfm @zhxchen17 @angelayi @suo @ydwu4 @Xia-Weiwen

Differential Revision: [D51078056](https://our.internmc.facebook.com/intern/diff/D51078056)

tugsbayasgalan committed:
[RELAND] Disallow skipping dynamo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[RELAND] Disallow skipping dynamo"


Previous discussion: https://github.com/pytorch/pytorch/pull/109476

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[RELAND] Disallow skipping dynamo"


Previous discussion: https://github.com/pytorch/pytorch/pull/109476

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zhxchen17 commented:
@pytorchbot merge

tugsbayasgalan committed:
Update on "[RELAND] Disallow skipping dynamo"


Previous discussion: https://github.com/pytorch/pytorch/pull/109476

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng avikchaudhuri gmagogsfm zhxchen17 angelayi Xia-Weiwen

[ghstack-poisoned]

tugsbayasgalan commented on torch/_export/exported_program.py:
```diff
@@ -238,6 +238,27 @@ def _construct_inp_pos_to_param_buffer_name(new_gm, graph_signature, state_dict)
+            equality_constraints=self.equality_constraints
+        )(*flat_args)
+
+    def __call__(self, *args, **kwargs):
```
Should we also do this for forward?

zhxchen17 commented:
Can you also support this for unflattened module? You can find it in .module(flat=False)

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51078056).

tugsbayasgalan committed:
Update on "[RELAND] Disallow skipping dynamo"


Previous discussion: https://github.com/pytorch/pytorch/pull/109476

In this PR, I made following additions to the original PR:
1) Unlifted graph module now runs the runtime assertions in its' forward call. 
2) When we retrace, we make sure we run the assertions to make sure user is tracing the module with correct inputs with respect to the assumptions we made during first tracing. The way I do is that I create new graph module type with modified call method. And the runtime assertions happen under torchdynamo.disable so that it is just run in eager directly. The reason is we don't this to be traced part of the graph. 
3) Both ep.module and capture_pre_autograd now returns _UnliftedGraphModule. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng avikchaudhuri gmagogsfm zhxchen17 angelayi Xia-Weiwen

Differential Revision: [D51078056](https://our.internmc.facebook.com/intern/diff/D51078056)

[ghstack-poisoned]

zhxchen17 commented:
@tugsbayasgalan test failures

zhxchen17 reviewed (changes requested):
@tugsbayasgalan test failures

tugsbayasgalan committed:
Update on "[RELAND] Disallow skipping dynamo"


Previous discussion: https://github.com/pytorch/pytorch/pull/109476

In this PR, I made following additions to the original PR:
1) Unlifted graph module now runs the runtime assertions in its' forward call. 
2) When we retrace, we make sure we run the assertions to make sure user is tracing the module with correct inputs with respect to the assumptions we made during first tracing. The way I do is that I create new graph module type with modified call method. And the runtime assertions happen under torchdynamo.disable so that it is just run in eager directly. The reason is we don't this to be traced part of the graph. 
3) Both ep.module and capture_pre_autograd now returns _UnliftedGraphModule. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng avikchaudhuri gmagogsfm zhxchen17 angelayi Xia-Weiwen

Differential Revision: [D51078056](https://our.internmc.facebook.com/intern/diff/D51078056)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[RELAND] Disallow skipping dynamo"


Previous discussion: https://github.com/pytorch/pytorch/pull/109476

In this PR, I made following additions to the original PR:
1) Unlifted graph module now runs the runtime assertions in its' forward call. 
2) When we retrace, we make sure we run the assertions to make sure user is tracing the module with correct inputs with respect to the assumptions we made during first tracing. The way I do is that I create new graph module type with modified call method. And the runtime assertions happen under torchdynamo.disable so that it is just run in eager directly. The reason is we don't this to be traced part of the graph. 
3) Both ep.module and capture_pre_autograd now returns _UnliftedGraphModule. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng avikchaudhuri gmagogsfm zhxchen17 angelayi Xia-Weiwen

Differential Revision: [D51078056](https://our.internmc.facebook.com/intern/diff/D51078056)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51078056).

zhxchen17 commented:
@pytorchbot merge
